pkg/types/v1/parameter.go

```
@@ -30,14 +32,81 @@ type (
 	// bytes, since this type will constantly be subject to such transformations.
 	Parameter struct {
 		// Name is the name of the parameter.
-		Name string
+		Name string `json:"name" yaml:"name"`
 		// Type is the parameter type.
-		Type ParameterType
+		Type ParameterType `json:"type" yaml:"type"`
 		// Value is the JSON encoded/decoded value of the parameter which is decoded based its Type.
-		Value any
+		Value any `json:"value" yaml:"value"`
 	}
 )
 
+func (p *Parameter) MarshalYAML() (interface{}, error) {
+	if err := p.Validate(); err != nil {
+		return nil, err
+	}
+
+	b, err := yaml.Marshal(struct {
+		Name  string
+		Type  ParameterType
+		Value any
+	}{
+		Name:  p.Name,
+		Type:  p.Type,
+		Value: p.Value,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not yaml marshal value: %w", err)
+	}
+
+	return b, nil
+}
+
+func (p *Parameter) UnmarshalYAML(value *yaml.Node) error {
+	if value == nil {
+		return nil
+	}
+
+	var (
+		partialParameter = &struct {
+			Name string
+			Type ParameterType
+		}{}
+	)
+
+	var err error
+	if err = value.Decode(&partialParameter); err != nil {
+		return errors.Errorf("failed to unmarshal parameter %s: %v", value.Value, err)
+	}
+
+	p.Name = partialParameter.Name
+	p.Type = partialParameter.Type
+
+	switch partialParameter.Type {
+	case ParameterTypeString, ParameterTypeConststring:
+		parameterValueStrPtr := &struct{ Value *string }{}
+		if err = value.Decode(&parameterValueStrPtr); err == nil {
```

I think it would make sense to set the parameter value to whatever is the value unmarshalled. if the value is a nil pointer, it's still type checked a nil string pointer that we can check when resolving the parameter overrides

---

Can you be clearer? This is mimicking what we do for the JSON unmarshaler as well

My understanding is that both pointers and values can be supplied with `ParameterTypeString` and `ParameterTypeConststring` and this takes care of it - you can check-out the tests in `parameter_test.go`

This also justifies the comment on line :93

---

as far as I understand the behaviour of the YAML unmarshaller, there are 3 cases that we need to be able to handle:
1. empty string
```yaml
value: ''
```

this is different than empty value, it is a literal empty string and the YAML unrmashaler would return a non-nil pointer pointing to an empty string, right? in which case you would get a nil err out of this line: `err = value.Decode(&parameterValueStrPtr)` and that branch would be executed

2. non-empty string -> the easiest one
again the you get a nil err out of that line of code and that branch gets executed

4. value is not present which is interpreted as a value that needs to be set explicitly by the user, otherwise the workflow launch should return an error. in this case the parameter would look like this:
```yaml
name: some-parameter
type: string
```

in this case the only way for us to figure out that the parameter is missing is to have nil string pointer and the YAML unmarshaler parsing this would set the value of the parameterValueStrPtr to nil, right? which again would mean that the following piece of code would not be executed

---

Took this one offline - the TL;DR is that we want to deal with potentially having nil values - by default we use `*string` and that's about it